File: hackerrank/practice/kruskalmstrsub.cpp
#include <iostream>
#include <map>
#include <algorithm>

using namespace std;

map<int, int> d;

class edge{
public:
	int node1, node2, weight;
	edge(){}
	edge(int n1, int n2, int w){
		node1 = n1;
		node2 = n2;
		weight = w;
	}
};

int find(int i){
	if(d[i]==-1)return i;
	return find(d[i]);
}


void Union(int x, int y){
	int xroot = find(x);
	int yroot = find(y);
	d[xroot] = yroot;
}

bool comp(edge e1, edge e2){
	if (e1.weight == e2.weight){
		return e1.node1+e1.node2+e1.weight < e2.node1+e2.node2+e2.weight ? true : false;
	}
	return e1.weight < e2.weight ? true : false;
}

int main(void){
	int m, n, n1, n2, w,start;
	cin >> n >> m;
	for (int i = 1; i <= n; i += 1) d[i] = -1;
	edge graph[m];
	for (int i = 0; i < m; i += 1){
		cin >> n1 >> n2 >> w;
		graph[i] = edge(n1, n2, w);
	}
	cin >> start;

	sort(graph, graph+m, comp);

	int ans = 0;
	for (edge e: graph){
		if (find(e.node1) == find(e.node2)){
			// Nothing to be done in that case. we just ignore the current edge
		}
		else{
			Union(e.node1, e.node2);
			ans += e.weight;
		}
	}
	cout << ans << endl;
}

File: hackerrank/practice/kruskalmstrsub.py
from collections import defaultdict
import sys
sys.setrecursionlimit(100000)

class Edge:
	def __init__(self, n1, n2, w):
		self.node1 = n1
		self.node2 = n2
		self.weight = w

	def __repr__(self):
		string = "%d %d %d"%(self.node1,self.node2,self.weight)
		return string


n, m = map(int, raw_input().split())
graph = []
for _ in xrange(m):
	n1, n2, w  = map(int, raw_input().split())
	graph.append(Edge(n1, n2, w))

raw_input() # starting point, still of no use

def comp(e1, e2):
	if e1.weight == e2.weight:
		return -1 if e1.node1+e1.node2+e1.weight < e2.node1+e2.node2+e2.weight else 0
	return -1 if e1.weight < e2.weight else 0

graph.sort(comp)
# print "Sorted graph:",graph

d = defaultdict(int)
for i in xrange(1,n+1):
	d[i] = -1
def insame(a, b):
	global d
	print "a:",a,"b:",b

	cur = d[a]
	while cur != -1:
		cur = d[cur]
		if cur == b:
			return True
	cur = d[b]
	while cur != -1:
		cur = d[cur]
		if cur == a:
			return True
	return False

def find(i):
	global d
	if d[i] == -1:
		return i
	return find(d[i])

def union(x, y):
	global d
	xroot = find(x)
	yroot = find(y)
	d[xroot] = yroot


ans = 0
for edge in graph:
	# print "current edge:",edge
	if find(edge.node1) == find(edge.node2):
		None
	else:
		union(edge.node1, edge.node2)
		ans += edge.weight
	# print "weight",ans
	# print "d:",d
print ans